gradio/analytics.py

```diff
@@ -0,0 +1,187 @@
+import warnings
+from distutils.version import StrictVersion
+from typing import Any
+
+import requests
+
+import gradio
+from gradio.context import Context
+from gradio.utils import GRADIO_VERSION
+
+ANALYTICS_URL = "https://api.gradio.app/"
+PKG_VERSION_URL = "https://api.gradio.app/pkg-version"
+
+
+def analytics_enabled() -> bool:
+    """
+    Returns: True if analytics are enabled, False otherwise.
+    """
+    return os.getenv("GRADIO_ANALYTICS_ENABLED", "True") == "True"
+
+
+def _do_analytics_request(url: str, data: dict[str, Any]) -> None:
+    try:
+        requests.post(url, data=data, timeout=5)
+    except (requests.ConnectionError, requests.exceptions.ReadTimeout):
+        pass  # do not push analytics if no network
+
+
+def version_check():
+    if not analytics_enabled():
+        return
+    try:
+        version_data = pkgutil.get_data(__name__, "version.txt")
+        if not version_data:
+            raise FileNotFoundError
+        current_pkg_version = version_data.decode("ascii").strip()
+        latest_pkg_version = requests.get(url=PKG_VERSION_URL, timeout=3).json()[
+            "version"
+        ]
+        if StrictVersion(latest_pkg_version) > StrictVersion(current_pkg_version):
+            print(
+                f"IMPORTANT: You are using gradio version {current_pkg_version}, "
+                f"however version {latest_pkg_version} is available, please upgrade."
+            )
+            print("--------")
+    except json.decoder.JSONDecodeError:
+        warnings.warn("unable to parse version details from package URL.")
+    except KeyError:
+        warnings.warn("package URL does not contain version info.")
+    except Exception:
+        pass
+
+
+def get_local_ip_address() -> str:
+    """
+    Gets the public IP address or returns the string "No internet connection" if unable
+    to obtain it or the string "Analytics disabled" if a user has disabled analytics.
+    Does not make a new request if the IP address has already been obtained in the
+    same Python session.
+    """
+    if not analytics_enabled():
+        return "Analytics disabled"
+
+    if Context.ip_address is None:
+        try:
+            ip_address = requests.get(
+                "https://checkip.amazonaws.com/", timeout=3
+            ).text.strip()
+        except (requests.ConnectionError, requests.exceptions.ReadTimeout):
+            ip_address = "No internet connection"
+        Context.ip_address = ip_address
+    else:
+        ip_address = Context.ip_address
+    return ip_address
+
+
+def initiated_analytics(data: dict[str, Any]) -> None:
+    if not analytics_enabled():
+        return
+
+    threading.Thread(
+        target=_do_analytics_request,
+        kwargs={
+            "url": f"{ANALYTICS_URL}gradio-initiated-analytics/",
+            "data": {**data, "ip_address": get_local_ip_address()},
+        },
+    ).start()
+
+
+def launched_telemetry(blocks: gradio.Blocks, data: dict[str, Any]) -> None:
```

Comment:
Can we rename this to `launched_analytics`

Reply:
Agreed, I'll make the change